File: services/worker/worker/core/git.py
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path
from urllib.parse import quote


class GitError(Exception):
    pass


@dataclass
class GitResult:
    ok: bool
    message: str


def _inject_token(url: str, token: str) -> str:
    if token and url.startswith("https://"):
        return url.replace("https://", f"https://x-access-token:{quote(token, safe='')}@", 1)
    return url


def _sanitize_stderr(stderr: str, token: str) -> str:
    if token:
        return stderr.replace(token, "***").replace(quote(token, safe=""), "***")
    return stderr


def clone_repo(repo_url: str, dest: Path, token: str = "", timeout: int = 300, branch: str = "") -> GitResult:
    try:
        dest.parent.mkdir(parents=True, exist_ok=True)
    except OSError as error:
        raise GitError(f"Could not create clone directory '{dest.parent}': {error}") from error

    command = ["git", "clone"]
    if branch:
        command.extend(["--branch", branch])
    command.extend([_inject_token(repo_url, token), str(dest)])

    try:
        result = subprocess.run(command, capture_output=True, text=True, timeout=timeout)
    except subprocess.TimeoutExpired as error:
        raise GitError("Timed out while cloning the repository") from error
    except Exception as error:
        raise GitError(f"Unexpected failure while cloning: {error}") from error

    if result.returncode != 0:
        raise GitError(_sanitize_stderr(result.stderr, token))
    return GitResult(ok=True, message="Repository cloned")


def pull_repo(repo_path: Path, token: str = "", timeout: int = 300, branch: str = "") -> GitResult:
    try:
        if branch:
            fetch_result = subprocess.run(
                ["git", "-C", str(repo_path), "fetch", "origin", f"+refs/heads/{branch}:refs/remotes/origin/{branch}"],
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            if fetch_result.returncode != 0:
                raise GitError(_sanitize_stderr(fetch_result.stderr, token))

            local_branch = subprocess.run(
                ["git", "-C", str(repo_path), "show-ref", "--verify", "--quiet", f"refs/heads/{branch}"],
                capture_output=True,
                text=True,
                timeout=timeout,
            )
            switch_command = (
                ["git", "-C", str(repo_path), "switch", branch]
                if local_branch.returncode == 0
                else ["git", "-C", str(repo_path), "switch", "--track", "-c", branch, f"origin/{branch}"]
            )
            switch_result = subprocess.run(switch_command, capture_output=True, text=True, timeout=timeout)
            if switch_result.returncode != 0:
                raise GitError(_sanitize_stderr(switch_result.stderr, token))

        result = subprocess.run(
            ["git", "-C", str(repo_path), "pull", "--ff-only"] + (["origin", branch] if branch else []),
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired as error:
        raise GitError("Timed out while updating the repository") from error
    except GitError:
        raise
    except Exception as error:
        raise GitError(f"Unexpected failure while updating: {error}") from error

    if result.returncode != 0:
        raise GitError(_sanitize_stderr(result.stderr, token))
    return GitResult(ok=True, message=f"Repository updated from branch '{branch}'" if branch else "Repository updated")


def list_remote_branches(repo_url: str, token: str = "", timeout: int = 30) -> list[str]:
    try:
        result = subprocess.run(
            ["git", "ls-remote", "--symref", _inject_token(repo_url, token), "HEAD", "refs/heads/*"],
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired as error:
        raise GitError("Timed out while reading remote branches") from error
    except Exception as error:
        raise GitError(f"Unexpected failure while reading remote branches: {error}") from error

    if result.returncode != 0:
        raise GitError(_sanitize_stderr(result.stderr, token))

    default_branch = ""
    branches: set[str] = set()
    for line in result.stdout.splitlines():
        if line.startswith("ref: refs/heads/") and line.endswith("\tHEAD"):
            default_branch = line.removeprefix("ref: refs/heads/").removesuffix("\tHEAD")
            continue
        _, separator, ref = line.partition("\t")
        if separator and ref.startswith("refs/heads/"):
            branches.add(ref.removeprefix("refs/heads/"))

    ordered = sorted(branches, key=str.casefold)
    if default_branch in branches:
        ordered.remove(default_branch)
        ordered.insert(0, default_branch)
    return ordered


def sync_repo(repo_url: str, repo_path: Path, token: str = "", timeout: int = 300, branch: str = "") -> GitResult:
    if (repo_path / ".git").is_dir():
        return pull_repo(repo_path, token=token, timeout=timeout, branch=branch)

    if repo_path.exists():
        if not repo_path.is_dir():
            raise GitError(f"Clone destination '{repo_path}' exists and is not a directory")
        try:
            next(repo_path.iterdir())
        except StopIteration:
            pass
        except OSError as error:
            raise GitError(f"Could not inspect clone destination '{repo_path}': {error}") from error
        else:
            raise GitError(f"Clone destination '{repo_path}' exists and is not an empty Git repository")

    return clone_repo(repo_url, repo_path, token=token, timeout=timeout, branch=branch)

File: services/worker-go/worker/core/git.go
package core

import (
	"fmt"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

type GitResult struct {
	Message string
}

func SyncRepo(repoURL string, dest string, token string, branch string) (GitResult, error) {
	if isGitRepo(dest) {
		return PullRepo(dest, token, branch)
	}
	if err := os.MkdirAll(filepath.Dir(dest), 0755); err != nil {
		return GitResult{}, fmt.Errorf("could not create clone directory '%s': %w", filepath.Dir(dest), err)
	}
	if info, err := os.Stat(dest); err == nil {
		if !info.IsDir() {
			return GitResult{}, fmt.Errorf("clone destination '%s' exists and is not a directory", dest)
		}
		entries, err := os.ReadDir(dest)
		if err != nil {
			return GitResult{}, fmt.Errorf("could not inspect clone destination '%s': %w", dest, err)
		}
		if len(entries) > 0 {
			return GitResult{}, fmt.Errorf("clone destination '%s' exists and is not an empty Git repository", dest)
		}
	}
	return CloneRepo(repoURL, dest, token, branch)
}

func CloneRepo(repoURL string, dest string, token string, branch string) (GitResult, error) {
	args := []string{"clone"}
	if strings.TrimSpace(branch) != "" {
		args = append(args, "--branch", strings.TrimSpace(branch))
	}
	args = append(args, injectToken(repoURL, token), dest)
	if _, err := gitOutput(300*time.Second, token, args...); err != nil {
		return GitResult{}, err
	}
	return GitResult{Message: "Repository cloned"}, nil
}

func PullRepo(repoPath string, token string, branch string) (GitResult, error) {
	branch = strings.TrimSpace(branch)
	if branch != "" {
		if _, err := gitOutput(300*time.Second, token, "-C", repoPath, "fetch", "origin", "+refs/heads/"+branch+":refs/remotes/origin/"+branch); err != nil {
			return GitResult{}, err
		}
		_, localErr := gitOutput(30*time.Second, token, "-C", repoPath, "show-ref", "--verify", "--quiet", "refs/heads/"+branch)
		switchArgs := []string{"-C", repoPath, "switch", branch}
		if localErr != nil {
			switchArgs = []string{"-C", repoPath, "switch", "--track", "-c", branch, "origin/" + branch}
		}
		if _, err := gitOutput(60*time.Second, token, switchArgs...); err != nil {
			return GitResult{}, err
		}
	}
	args := []string{"-C", repoPath, "pull", "--ff-only"}
	if branch != "" {
		args = append(args, "origin", branch)
	}
	if _, err := gitOutput(300*time.Second, token, args...); err != nil {
		return GitResult{}, err
	}
	if branch != "" {
		return GitResult{Message: "Repository updated from branch '" + branch + "'"}, nil
	}
	return GitResult{Message: "Repository updated"}, nil
}

func ListRemoteBranches(repoURL string, token string) ([]string, error) {
	output, err := gitOutput(30*time.Second, token, "ls-remote", "--symref", injectToken(repoURL, token), "HEAD", "refs/heads/*")
	if err != nil {
		return nil, err
	}
	defaultBranch := ""
	branches := map[string]bool{}
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "ref: refs/heads/") && strings.HasSuffix(line, "\tHEAD") {
			defaultBranch = strings.TrimSuffix(strings.TrimPrefix(line, "ref: refs/heads/"), "\tHEAD")
			continue
		}
		parts := strings.SplitN(line, "\t", 2)
		if len(parts) == 2 && strings.HasPrefix(parts[1], "refs/heads/") {
			branches[strings.TrimPrefix(parts[1], "refs/heads/")] = true
		}
	}
	ordered := make([]string, 0, len(branches))
	for branch := range branches {
		ordered = append(ordered, branch)
	}
	sort.Slice(ordered, func(i, j int) bool {
		return strings.ToLower(ordered[i]) < strings.ToLower(ordered[j])
	})
	if defaultBranch != "" {
		for index, branch := range ordered {
			if branch == defaultBranch {
				ordered = append(ordered[:index], ordered[index+1:]...)
				ordered = append([]string{defaultBranch}, ordered...)
				break
			}
		}
	}
	return ordered, nil
}

func isGitRepo(path string) bool {
	info, err := os.Stat(filepath.Join(path, ".git"))
	return err == nil && info.IsDir()
}

func injectToken(repoURL string, token string) string {
	if strings.TrimSpace(token) == "" || !strings.HasPrefix(repoURL, "https://") {
		return repoURL
	}
	return strings.Replace(repoURL, "https://", "https://x-access-token:"+url.QueryEscape(token)+"@", 1)
}

func gitOutput(timeout time.Duration, token string, args ...string) (string, error) {
	ctxDone := time.After(timeout)
	cmd := exec.Command("git", args...)
	outputCh := make(chan struct {
		text string
		err  error
	}, 1)
	go func() {
		output, err := cmd.CombinedOutput()
		outputCh <- struct {
			text string
			err  error
		}{text: strings.TrimSpace(string(output)), err: err}
	}()
	select {
	case result := <-outputCh:
		if result.err != nil {
			return "", fmt.Errorf("%s", sanitizeOutput(result.text, token))
		}
		return result.text, nil
	case <-ctxDone:
		if cmd.Process != nil {
			_ = cmd.Process.Kill()
		}
		return "", fmt.Errorf("timed out while running git")
	}
}

func sanitizeOutput(output string, token string) string {
	if strings.TrimSpace(token) == "" {
		return output
	}
	return strings.ReplaceAll(strings.ReplaceAll(output, token, "***"), url.QueryEscape(token), "***")
}
